packages/web-components/src/dropdown/dropdown.ts

```diff
@@ -0,0 +1,1044 @@
+    if (this.$fastController.isConnected) {
+      if (this.disabled || !this.required) {
+        this.elementInternals.setValidity({});
+        return;
+      }
+
+      const valueMissing = this.required && this.listbox.selectedOptions.length === 0;
+
+      this.elementInternals.setValidity(
+        { valueMissing, ...flags },
+        message ?? this.validationMessage,
+        anchor ?? this.listbox.enabledOptions[0],
+      );
+    }
+  }
+
+  protected updateFreeformOption(value: string = this.control.value): void {
+    if (!this.freeformOption) {
+      return;
+    }
+
+    if (
+      value === '' ||
+      this.filterOptions(
+        value,
+        this.enabledOptions.filter(x => !x.freeform),
+      ).length
+    ) {
+      this.freeformOption.value = '';
+      this.freeformOption.selected = false;
+      this.freeformOption.hidden = true;
+      return;
+    }
+
+    this.freeformOption.value = value;
+    this.freeformOption.hidden = false;
+  }
+}
+
+export class Dropdown extends BaseDropdown {
+  /**
+   * Static property for the anchor positioning fallback observer. The observer is used to flip the listbox when it is
+   * out of view.
+   *
+   * @remarks This is only used when the browser does not support CSS anchor positioning, and the CSS anchor polyfill is
+   * not present.
+   *
+   * @internal
+   */
+  private static AnchorPositionFallbackObserver: IntersectionObserver;
+
+  /**
+   * Fallback style element for anchor positioning.
+   *
+   * @remarks This is only used when the browser does not support CSS anchor positioning, and the CSS anchor polyfill is
+   * present.
+   *
+   * @internal
+   */
+  private anchorPositionFallbackStyleElement: HTMLStyleElement | null = null;
+
+  /**
+   * Static property for the anchor positioning fallback style elements.
+   *
+   * @remarks This is only used when the browser does not support CSS anchor positioning, and the CSS anchor polyfill is
+   * present.
+   *
+   * @internal
+   */
+  private static AnchorPositionFallbackStyleElements: Map<BaseDropdown, HTMLStyleElement> = new Map();
+
+  /**
+   * Static property for the anchor positioning fallback timeout reference. This is used to prevent multiple idle
+   * callbacks from being scheduled when multiple dropdowns are created in quick succession.
+   *
+   * @remarks This is only used when the browser does not support CSS anchor positioning, and the CSS anchor polyfill is
+   * present.
+   *
+   * @internal
+   */
+  private static AnchorPositionFallbackTimeout: number | null;
+
+  /**
+   * Applies the `--listbox-max-height` style to the listbox based on the dropdown's position in the viewport. This
+   * prevents the listbox from growing beyond the available space.
+   *
+   * @internal
+   */
+  private scrollWindowHandler = () => {
+    const rect = this.getBoundingClientRect();
+    if (rect.top < window.innerHeight - rect.bottom) {
```

Comment:
`getBoundingClientRect()` and `window.innerHeight` both trigger a forced layout reflow which could have some perf problems inside a window.onscroll handler. At a minimum we'd want to debounce this. At a maximum, we'd want to figure out an alternative? like max-height: 40vh or something?

https://gist.github.com/paulirish/5d52fb081b3570c81e3a

Reply:
I removed this event listener and opted for the `max-height: 40vh` route.